### ISSUE TITLE
Any intentions of R interface?

### ISSUE THREAD

**Issue by MansMeg** [tensorflow/tensorflow#387]
Hi!

Im working with the python wrapper for tensorflow. Is there any intentions of implement similar wrappers for the R languages?

**Comment by thirdwing** [comment]
There is some attempt, like https://github.com/terrytangyuan/rflow

It shouldn't be hard to have an R wrapper like the python one, but it will be quite time-consuming.

**Comment by MansMeg** [comment]
Thanks! I looked it up and it still depends quite much on Python. 

Why would it be time consuming, is TensorFlow built in Python or is mainly built in C++?

**Comment by thirdwing** [comment]
@MansMeg For R interface, there are two possible solutions:

(1) SWIG, used in shogun (https://github.com/shogun-toolbox/shogun)

(2) Rcpp, used in mxnet (https://github.com/dmlc/mxnet/tree/master/R-package), which is also a deep learning platform

As I understand, the core part is in C++ and that's why I said it shouldn't be hard. I am part of Rcpp team, so I know it is quite easy to call C++ from R.

But there is much more work between calling tensorflow from R and building a useful interface, right?

**Comment by MansMeg** [comment]
Of course, there is some effort in building an interface! I just think that it is a difference if the core also would be built in Python, then it would take much more time (and maintenance) to have an R interface - porting code and keeping it updated.

Thanks for the links!

**Comment by jjallaire** [comment]
@thirdwing I'm also quite interested in R bindings. I think that mapping directly from the C++ API to R via either SWIG or Rcpp is a necessary but not sufficient condition to create really compelling R bindings. What we really need are idiomatic bindings that take advantage of R formulas, do automatic differentiation, etc. I think it's correct that it would take time and maintenance to do this properly but the result could be really stunning!

BTW, another possible path is alluded to here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/public/tensor_c_api.h#L55-L59

In short, it's suggested that some languages may choose to bypass the C++/SWIG route entirely and just generate the requisite ProtoBuf input to the C API. It's not clear to me whether this would be a wasteful duplication of the existing C++ layer or something that would actually simplify creating idiomatic R bindings.

**Comment by thirdwing** [comment]
Thinking about automatic differentiation, it could be quite really interesting if we can use R formulas. I will look into this.

**Comment by jjallaire** [comment]
@girving Interested in your guidance on the various ways to pursue R bindings. Read your comment here providing the lay of the land for Rust (https://github.com/tensorflow/tensorflow/issues/388#issuecomment-161019498). We obviously can pursue SWIG bindings to the existing C++ classes and then pickup additional functionality from the C++ layer once more of the features from the python layer are moved there.

However, we're most interested in creating idiomatic bindings for R and to get this exactly right are in no way deterred by the fact that it may take a lot of effort (much of which is duplicated). I can think of a few ways to approach creating these bindings:

1. Create a wrapper for sessions / invocation using the C API then create the graph definition bindings by going directly to proto. This possibility seems to be anticipated/encouraged here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/public/tensor_c_api.h#L55-L59

2. Call the C++ layer via SWIG bindings.

3. Call python directly from R using e.g. boost::python or pybind11. Assuming that NumPy arrays can be marshaled to from R arrays with no copying I'm assuming this would have no performance problems since the python code is just defining a graph.

To me the first option has the greatest appeal because there would be no impedance problems associated with translating idiomatic R into graphs, we'd just figure out the right syntax and express it as a graph definition rather than contorting it through the C++ layer. The comment above seems to imply that this would be an effective path, but I haven't spent enough time looking to know whether this would be a monumental amount of work that would be constantly challenged to keep up or just something to grind through once (not deterred at all by having to spend on the order of hundreds of hours in the initial effort). I also like the idea of using the C API because those entry points could be loaded dynamically, meaning that the R bindings could be built separately and made available on CRAN and could work with any installed version of TensorFlow.

We could also combine the creation of proto-based idiomatic R bindings with another binding that parrots the python API. This would allow R users to easily translate and use python or C++ API based examples but at the same time have a first class R binding that makes maximum sense to R users.

Your thoughts and feedback very much appreciated!

**Comment by martinwicke** [comment]
A lot of the functionality of tensorflow is written in python, and will
migrate only slowly down to C++.

The GraphDef proto should only change slowly (and with the best guarantees
for compatibility), so writing protos directly will require the least
velocity of change to keep things working. If your main goal is to run
existing graphs from R (that's unlikely to be the case, but that's the
situation for node.js or Java where the main focus is on running
pre-trained models), then this would be sufficient.

On the other hand, the API will be significantly less powerful if you cut
away the python layer entirely. Mainly this is because the gradients and
shape inference layers are written in python, but also some ops are python
functions that combine kernels into more complex ops (image_ops is a good
example). This makes #3 look much better than it otherwise would, even
though the API is changing faster than the GraphDef is.

**Comment by jjallaire** [comment]
@martinwicke Thanks for the follow up. So it sounds like calling directly from R to python might be the best chance of providing a robust and complete R interface. We could first create a variation of the R interface that is entirely syntax compatible with the python interface then incrementally introduce various types of R sugar to make things more natural and compact for R users.

In order to stay in sync this would ideally take advantage of some machine readable specification of the python interface and documentation. What do you recommend as the best course here?

**Comment by martinwicke** [comment]
We have the op registry, but it only contains the C++ api, as well as the functions for gradients and shape inference. The closest we have for the python layer is the API docs, which is generated by scraping the source code, and therefore should always be up to date. This quickly turns icky, but you could actually get a reasonable snapshot by going through the python doc .md files and collecting headlines.

**Comment by refack** [comment]
Has there been any progress? I had a few nudges from R users to throw my hat in the ring...

**Comment by jjallaire** [comment]
I've got an R package that has a binding to the TensorFlow C-interface here: https://github.com/jjallaire/TensorFlow

The main trick of this package is that it doesn't actually build against the TF library (which will be impossible on CRAN unless CRAN decides to install TensorFlow on their build machines) but rather dynamically discovers it's location and loads up the symbols manually.  

In principle this is the first step of being able to run TF models defined as protobufs. However, as alluded to previously in this thread I think a more fruitful direction is going to be just calling all the Python code directly from within R. I'm not sure I'll have time to explore this soon but would certainly try to help out with any other efforts to make this happen.

**Comment by rajshah4** [comment]
I was able to make a Shiny/R app that interfaces with tensorflow using rPython.  https://github.com/rajshah4/tensorflow_shiny  
Awkward, but it worked.  Hoping that a better solution gets developed.

**Comment by aselle** [comment]
Any objections to closing this given that there are now relatively working R bindings https://github.com/rstudio/tensorflow

**Comment by MansMeg** [comment]
No, great!

Skickat från min iPhone
